File: MainWindowTabKaraokeSync.vb
Imports System.Text
Imports System.Windows
Imports Subtitle_draft_GMTPC.Services

Partial Class MainWindow

#Region "Karaoke Sync - Fields"

    Private _isSyncUpdating As Boolean = False

#End Region

#Region "Karaoke Sync - Event Handlers"

    ''' <summary>
    ''' Khi input thay đổi, cập nhật số dòng
    ''' </summary>
    Private Sub TxtSyncInput_TextChanged(sender As Object, e As TextChangedEventArgs)
        If _isSyncUpdating Then Return
        Try
            _isSyncUpdating = True
            Dim content = TxtSyncInput.Text.Trim()
            If String.IsNullOrWhiteSpace(content) Then
                TxtSyncInputCount.Text = ""
                TxtSyncOutput.Text = ""
                TxtSyncOutputCount.Text = ""
                Return
            End If

            Dim lines = content.Split({Environment.NewLine, vbCr, vbLf}, StringSplitOptions.RemoveEmptyEntries)
            TxtSyncInputCount.Text = String.Format("({0} dòng)", lines.Length)

            ' Tự động sync khi có input
            SyncTimeCodes()
        Catch ex As Exception
            TxtSyncInputCount.Text = String.Format("(Lỗi: {0})", ex.Message)
        Finally
            _isSyncUpdating = False
        End Try
    End Sub

    ''' <summary>
    ''' Khi time input thay đổi, auto sync
    ''' </summary>
    Private Sub TxtSyncTimeInput_TextChanged(sender As Object, e As TextChangedEventArgs)
        If _isSyncUpdating Then Return
        Try
            SyncTimeCodes()
        Catch ex As Exception
            ' Ignore parse errors during typing
        End Try
    End Sub

    ''' <summary>
    ''' Offset toàn bộ time code theo desired start time (giữ nguyên duration)
    ''' </summary>
    Private Sub BtnSyncTime_Click(sender As Object, e As RoutedEventArgs)
        Try
            SyncTimeCodes()
        Catch ex As Exception
            MessageBox.Show("Lỗi sync time: " & ex.Message, "Lỗi", MessageBoxButton.OK, MessageBoxImage.Error)
        End Try
    End Sub

    ''' <summary>
    ''' Copy kết quả sync
    ''' </summary>
    Private Sub BtnCopySync_Click(sender As Object, e As RoutedEventArgs)
        If String.IsNullOrWhiteSpace(TxtSyncOutput.Text) Then Return
        Try
            Clipboard.SetText(TxtSyncOutput.Text)
            ShowToastSync("📋 Đã copy kết quả sync!")
        Catch ex As Exception
            MessageBox.Show("Lỗi: " & ex.Message, "Lỗi", MessageBoxButton.OK, MessageBoxImage.Error)
        End Try
    End Sub

#End Region

#Region "Karaoke Sync - Sync Logic"

    ''' <summary>
    ''' Thực hiện sync time codes
    ''' Logic: Tính offset từ dòng đầu tiên, áp dụng offset đó cho tất cả các dòng
    ''' </summary>
    Private Sub SyncTimeCodes()
        Dim inputContent = TxtSyncInput.Text.Trim()
        If String.IsNullOrWhiteSpace(inputContent) Then
            TxtSyncOutput.Text = ""
            TxtSyncOutputCount.Text = ""
            Return
        End If

        ' Parse desired start time từ textbox (format: HH:MM:SS.mmm hoặc HH:MM:SS,mmm)
        Dim desiredStartMs As Long = ParseTimeInputToMs(TxtSyncTimeInput.Text)
        If desiredStartMs < 0 Then
            TxtSyncOutput.Text = inputContent
            TxtSyncOutputCount.Text = "(Lỗi: Time format không hợp lệ)"
            Return
        End If

        Dim lines = inputContent.Split({Environment.NewLine, vbCr, vbLf}, StringSplitOptions.RemoveEmptyEntries)

        ' Tìm dòng đầu tiên có time code (Dialogue hoặc Comment)
        Dim firstLineStartTimeMs As Long? = Nothing
        Dim offsetMs As Long = 0

        For Each line In lines
            ' Thử parse SRT format
            Dim srtMatch = System.Text.RegularExpressions.Regex.Match(line, "(\d{2}:\d{2}:\d{2},\d{3})\s*-->\s*(\d{2}:\d{2}:\d{2},\d{3})")
            If srtMatch.Success Then
                Dim startTimeStr = srtMatch.Groups(1).Value
                firstLineStartTimeMs = ParseSrtTimeToMs(startTimeStr)
                Exit For
            End If

            ' Thử parse ASS format (Dialogue hoặc Comment)
            Dim assMatch = System.Text.RegularExpressions.Regex.Match(line, "^(?:Dialogue|Comment):\s*\d+,(\d+:\d{2}:\d{2}\.\d{2}),(\d+:\d{2}:\d{2}\.\d{2})", System.Text.RegularExpressions.RegexOptions.IgnoreCase)
            If assMatch.Success Then
                Dim startTimeStr = assMatch.Groups(1).Value
                firstLineStartTimeMs = ParseAssTimeToMs(startTimeStr)
                Exit For
            End If
        Next

        ' Nếu không tìm thấy time code nào
        If firstLineStartTimeMs Is Nothing Then
            TxtSyncOutput.Text = inputContent
            TxtSyncOutputCount.Text = "(0 dòng đã sync)"
            Return
        End If

        ' Tính offset: desired - actual
        offsetMs = desiredStartMs - firstLineStartTimeMs

        ' Áp dụng offset cho tất cả các dòng
        Dim sb = New StringBuilder()
        Dim processedCount As Integer = 0

        For Each line In lines
            Dim processedLine = line

            ' Xử lý SRT format: 00:00:00,000 --> 00:00:00,000
            Dim srtMatch = System.Text.RegularExpressions.Regex.Match(line, "(\d{2}:\d{2}:\d{2},\d{3})\s*-->\s*(\d{2}:\d{2}:\d{2},\d{3})")
            If srtMatch.Success Then
                Dim startTimeStr = srtMatch.Groups(1).Value
                Dim endTimeStr = srtMatch.Groups(2).Value

                Dim startMs = ParseSrtTimeToMs(startTimeStr)
                Dim endMs = ParseSrtTimeToMs(endTimeStr)

                ' Offset time, đảm bảo không âm
                Dim newStartMs = Math.Max(0L, startMs + offsetMs)
                Dim newEndMs = Math.Max(0L, endMs + offsetMs)

                ' Giữ nguyên duration
                Dim duration = endMs - startMs
                newEndMs = newStartMs + duration

                Dim newStartStr = MsToSrtTime(newStartMs)
                Dim newEndStr = MsToSrtTime(newEndMs)

                processedLine = line.Replace(startTimeStr, newStartStr).Replace(endTimeStr, newEndStr)
                processedCount += 1
            Else
                ' Xử lý ASS format: Dialogue hoặc Comment
                Dim assMatch = System.Text.RegularExpressions.Regex.Match(line, "^((?:Dialogue|Comment):\s*\d+,)(\d+:\d{2}:\d{2}\.\d{2}),(\d+:\d{2}:\d{2}\.\d{2})(.*)$", System.Text.RegularExpressions.RegexOptions.IgnoreCase)
                If assMatch.Success Then
                    Dim prefix = assMatch.Groups(1).Value
                    Dim startTimeStr = assMatch.Groups(2).Value
                    Dim endTimeStr = assMatch.Groups(3).Value
                    Dim suffix = assMatch.Groups(4).Value

                    Dim startMs = ParseAssTimeToMs(startTimeStr)
                    Dim endMs = ParseAssTimeToMs(endTimeStr)

                    ' Offset time, đảm bảo không âm
                    Dim newStartMs = Math.Max(0L, startMs + offsetMs)
                    Dim newEndMs = Math.Max(0L, endMs + offsetMs)

                    ' Giữ nguyên duration
                    Dim duration = endMs - startMs
                    newEndMs = newStartMs + duration

                    Dim newStartStr = MsToAssTime(newStartMs)
                    Dim newEndStr = MsToAssTime(newEndMs)

                    processedLine = prefix & newStartStr & "," & newEndStr & suffix
                    processedCount += 1
                End If
            End If

            sb.AppendLine(processedLine)
        Next

        TxtSyncOutput.Text = sb.ToString().TrimEnd()
        TxtSyncOutputCount.Text = String.Format("({0} dòng đã sync)", processedCount)
    End Sub

#End Region

#Region "Karaoke Sync - Time Parsing Helpers"

    ''' <summary>
    ''' Parse time input string (HH:MM:SS.mmm hoặc HH:MM:SS,mmm) sang milliseconds
    ''' Trả về -1 nếu format không hợp lệ
    ''' </summary>
    Private Function ParseTimeInputToMs(timeStr As String) As Long
        If String.IsNullOrWhiteSpace(timeStr) Then Return 0L

        ' Chuẩn hóa: thay comma bằng dot để统一 xử lý
        timeStr = timeStr.Trim().Replace(","c, "."c)

        ' Regex: HH:MM:SS.mmm (có thể bỏ HH hoặc MM)
        Dim match = System.Text.RegularExpressions.Regex.Match(timeStr, "^(\d+):(\d{2}):(\d{2})\.(\d+)$")
        If Not match.Success Then
            Return -1L
        End If

        Try
            Dim hours = Integer.Parse(match.Groups(1).Value)
            Dim minutes = Integer.Parse(match.Groups(2).Value)
            Dim seconds = Integer.Parse(match.Groups(3).Value)
            Dim fracSeconds = match.Groups(4).Value

            ' Parse fractional seconds (có thể là 1-3 digits)
            Dim milliseconds As Integer = 0
            If fracSeconds.Length = 1 Then
                milliseconds = Integer.Parse(fracSeconds) * 100
            ElseIf fracSeconds.Length = 2 Then
                milliseconds = Integer.Parse(fracSeconds) * 10
            Else
                milliseconds = Integer.Parse(fracSeconds.Substring(0, 3))
            End If

            Return (hours * 3600L + minutes * 60L + seconds) * 1000L + milliseconds
        Catch
            Return -1L
        End Try
    End Function

    ''' <summary>
    ''' Parse SRT time string (00:00:00,000) sang milliseconds
    ''' </summary>
    Private Function ParseSrtTimeToMs(timeStr As String) As Long
        Dim parts = timeStr.Split(":"c, ","c)
        If parts.Length <> 4 Then Return 0L

        Dim hours = Integer.Parse(parts(0))
        Dim minutes = Integer.Parse(parts(1))
        Dim seconds = Integer.Parse(parts(2))
        Dim milliseconds = Integer.Parse(parts(3))

        Return hours * 3600000L + minutes * 60000L + seconds * 1000L + milliseconds
    End Function

    ''' <summary>
    ''' Convert milliseconds sang SRT time string (00:00:00,000)
    ''' </summary>
    Private Function MsToSrtTime(ms As Long) As String
        Dim isNegative = ms < 0
        ms = Math.Abs(ms)

        Dim hours = ms \ 3600000L
        ms = ms Mod 3600000L
        Dim minutes = ms \ 60000L
        ms = ms Mod 60000L
        Dim seconds = ms \ 1000L
        Dim milliseconds = ms Mod 1000L

        Return String.Format("{0:D2}:{1:D2}:{2:D2},{3:D3}", hours, minutes, seconds, milliseconds)
    End Function

    ''' <summary>
    ''' Parse ASS time string (H:MM:SS.cc) sang milliseconds
    ''' </summary>
    Private Function ParseAssTimeToMs(timeStr As String) As Long
        Dim parts = timeStr.Split(":"c, "."c)
        If parts.Length <> 4 Then Return 0L

        Dim hours = Integer.Parse(parts(0))
        Dim minutes = Integer.Parse(parts(1))
        Dim seconds = Integer.Parse(parts(2))
        Dim centiseconds = Integer.Parse(parts(3))

        Return hours * 3600000L + minutes * 60000L + seconds * 1000L + centiseconds * 10L
    End Function

    ''' <summary>
    ''' Convert milliseconds sang ASS time string (H:MM:SS.cc)
    ''' </summary>
    Private Function MsToAssTime(ms As Long) As String
        Dim isNegative = ms < 0
        ms = Math.Abs(ms)

        Dim hours = ms \ 3600000L
        ms = ms Mod 3600000L
        Dim minutes = ms \ 60000L
        ms = ms Mod 60000L
        Dim seconds = ms \ 1000L
        Dim centiseconds = (ms Mod 1000L) \ 10L

        Return String.Format("{0}:{1:D2}:{2:D2}.{3:D2}", hours, minutes, seconds, centiseconds)
    End Function

#End Region

#Region "Karaoke Sync - Toast"

    Private Async Sub ShowToastSync(message As String)
        ToastTextSync.Text = message
        ToastBorderSync.Visibility = Visibility.Visible
        Await Task.Delay(2000)
        ToastBorderSync.Visibility = Visibility.Collapsed
    End Sub

#End Region

End Class

File: MainWindowTabKaraokeSync.cs
using System;
using System.Collections.Generic;
using System.Windows;
using System.Windows.Controls;
using System.Threading.Tasks;
using System.Text;
using Subtitle_draft_GMTPC.Models;
using Subtitle_draft_GMTPC.Services;

namespace Subtitle_draft_GMTPC
{
    public partial class MainWindow : Window
    {

        #region Karaoke Sync - Fields

        private bool _isSyncUpdating = false;

        #endregion

        #region Karaoke Sync - Event Handlers

        /// <summary>
        /// Khi input thay đổi, cập nhật số dòng
        /// </summary>
        private void TxtSyncInput_TextChanged(object sender, TextChangedEventArgs e)
        {
            if (_isSyncUpdating) return;
            try
            {
                _isSyncUpdating = true;
                var content = TxtSyncInput.Text.Trim();
                if (string.IsNullOrWhiteSpace(content))
                {
                    TxtSyncInputCount.Text = "";
                    TxtSyncOutput.Text = "";
                    TxtSyncOutputCount.Text = "";
                    return;
                }

                var lines = content.Split(new[] { Environment.NewLine, "\r", "\n" }, StringSplitOptions.RemoveEmptyEntries);
                TxtSyncInputCount.Text = string.Format("({0} dòng)", lines.Length);

                // Tự động sync khi có input
                SyncTimeCodes();
            }
            catch (Exception ex)
            {
                TxtSyncInputCount.Text = string.Format("(Lỗi: {0})", ex.Message);
            }
            finally
            {
                _isSyncUpdating = false;
            }
        }

        /// <summary>
        /// Khi time input thay đổi, auto sync
        /// </summary>
        private void TxtSyncTimeInput_TextChanged(object sender, TextChangedEventArgs e)
        {
            if (_isSyncUpdating) return;
            try
            {
                SyncTimeCodes();
            }
            catch (Exception)
            {
                // Ignore parse errors during typing
            }
        }

        /// <summary>
        /// Offset toàn bộ time code theo desired start time (giữ nguyên duration)
        /// </summary>
        private void BtnSyncTime_Click(object sender, RoutedEventArgs e)
        {
            try
            {
                SyncTimeCodes();
            }
            catch (Exception ex)
            {
                System.Windows.MessageBox.Show("Lỗi sync time: " + ex.Message, "Lỗi", System.Windows.MessageBoxButton.OK, System.Windows.MessageBoxImage.Error);
            }
        }

        /// <summary>
        /// Copy kết quả sync
        /// </summary>
        private void BtnCopySync_Click(object sender, RoutedEventArgs e)
        {
            if (string.IsNullOrWhiteSpace(TxtSyncOutput.Text)) return;
            try
            {
                Clipboard.SetText(TxtSyncOutput.Text);
                ShowToastSync("📋 Đã copy kết quả sync!");
            }
            catch (Exception ex)
            {
                System.Windows.MessageBox.Show("Lỗi: " + ex.Message, "Lỗi", System.Windows.MessageBoxButton.OK, System.Windows.MessageBoxImage.Error);
            }
        }

        #endregion

        #region Karaoke Sync - Sync Logic

        /// <summary>
        /// Thực hiện sync time codes
        /// Logic: Tính offset từ dòng đầu tiên, áp dụng offset đó cho tất cả các dòng
        /// </summary>
        private void SyncTimeCodes()
        {
            var inputContent = TxtSyncInput.Text.Trim();
            if (string.IsNullOrWhiteSpace(inputContent))
            {
                TxtSyncOutput.Text = "";
                TxtSyncOutputCount.Text = "";
                return;
            }

            // Parse desired start time từ textbox (format: HH:MM:SS.mmm hoặc HH:MM:SS,mmm)
            var desiredStartMs = ParseTimeInputToMs(TxtSyncTimeInput.Text);
            if (desiredStartMs < 0)
            {
                TxtSyncOutput.Text = inputContent;
                TxtSyncOutputCount.Text = "(Lỗi: Time format không hợp lệ)";
                return;
            }

            var lines = inputContent.Split(new[] { Environment.NewLine, "\r", "\n" }, StringSplitOptions.RemoveEmptyEntries);

            // Tìm dòng đầu tiên có time code (Dialogue hoặc Comment)
            long? firstLineStartTimeMs = null;
            long offsetMs = 0;

            foreach (var line in lines)
            {
                // Thử parse SRT format
                var srtMatch = System.Text.RegularExpressions.Regex.Match(line, @"(\d{2}:\d{2}:\d{2},\d{3})\s*-->\s*(\d{2}:\d{2}:\d{2},\d{3})");
                if (srtMatch.Success)
                {
                    var startTimeStr = srtMatch.Groups[1].Value;
                    firstLineStartTimeMs = ParseSrtTimeToMs(startTimeStr);
                    break;
                }

                // Thử parse ASS format (Dialogue hoặc Comment)
                var assMatch = System.Text.RegularExpressions.Regex.Match(line, @"^(?:Dialogue|Comment):\s*\d+,(\d+:\d{2}:\d{2}\.\d{2}),(\d+:\d{2}:\d{2}\.\d{2})", System.Text.RegularExpressions.RegexOptions.IgnoreCase);
                if (assMatch.Success)
                {
                    var startTimeStr = assMatch.Groups[1].Value;
                    firstLineStartTimeMs = ParseAssTimeToMs(startTimeStr);
                    break;
                }
            }

            // Nếu không tìm thấy time code nào
            if (firstLineStartTimeMs == null)
            {
                TxtSyncOutput.Text = inputContent;
                TxtSyncOutputCount.Text = "(0 dòng đã sync)";
                return;
            }

            // Tính offset: desired - actual
            offsetMs = desiredStartMs - firstLineStartTimeMs.Value;

            // Áp dụng offset cho tất cả các dòng
            var sb = new StringBuilder();
            int processedCount = 0;

            foreach (var line in lines)
            {
                var processedLine = line;

                // Xử lý SRT format: 00:00:00,000 --> 00:00:00,000
                var srtMatch = System.Text.RegularExpressions.Regex.Match(line, @"(\d{2}:\d{2}:\d{2},\d{3})\s*-->\s*(\d{2}:\d{2}:\d{2},\d{3})");
                if (srtMatch.Success)
                {
                    var startTimeStr = srtMatch.Groups[1].Value;
                    var endTimeStr = srtMatch.Groups[2].Value;

                    var startMs = ParseSrtTimeToMs(startTimeStr);
                    var endMs = ParseSrtTimeToMs(endTimeStr);

                    // Offset time, đảm bảo không âm
                    var newStartMs = Math.Max(0L, startMs + offsetMs);
                    var newEndMs = Math.Max(0L, endMs + offsetMs);

                    // Giữ nguyên duration
                    var duration = endMs - startMs;
                    newEndMs = newStartMs + duration;

                    var newStartStr = MsToSrtTime(newStartMs);
                    var newEndStr = MsToSrtTime(newEndMs);

                    processedLine = line.Replace(startTimeStr, newStartStr).Replace(endTimeStr, newEndStr);
                    processedCount += 1;
                }
                else
                {
                    // Xử lý ASS format: Dialogue hoặc Comment
                    var assMatch = System.Text.RegularExpressions.Regex.Match(line, @"^((?:Dialogue|Comment):\s*\d+,)(\d+:\d{2}:\d{2}\.\d{2}),(\d+:\d{2}:\d{2}\.\d{2})(.*)$", System.Text.RegularExpressions.RegexOptions.IgnoreCase);
                    if (assMatch.Success)
                    {
                        var prefix = assMatch.Groups[1].Value;
                        var startTimeStr = assMatch.Groups[2].Value;
                        var endTimeStr = assMatch.Groups[3].Value;
                        var suffix = assMatch.Groups[4].Value;

                        var startMs = ParseAssTimeToMs(startTimeStr);
                        var endMs = ParseAssTimeToMs(endTimeStr);

                        // Offset time, đảm bảo không âm
                        var newStartMs = Math.Max(0L, startMs + offsetMs);
                        var newEndMs = Math.Max(0L, endMs + offsetMs);

                        // Giữ nguyên duration
                        var duration = endMs - startMs;
                        newEndMs = newStartMs + duration;

                        var newStartStr = MsToAssTime(newStartMs);
                        var newEndStr = MsToAssTime(newEndMs);

                        processedLine = prefix + newStartStr + "," + newEndStr + suffix;
                        processedCount += 1;
                    }
                }

                sb.AppendLine(processedLine);
            }

            TxtSyncOutput.Text = sb.ToString().TrimEnd();
            TxtSyncOutputCount.Text = string.Format("({0} dòng đã sync)", processedCount);
        }

        #endregion

        #region Karaoke Sync - Time Parsing Helpers

        /// <summary>
        /// Parse time input string (HH:MM:SS.mmm hoặc HH:MM:SS,mmm) sang milliseconds
        /// Trả về -1 nếu format không hợp lệ
        /// </summary>
        private long ParseTimeInputToMs(string timeStr)
        {
            if (string.IsNullOrWhiteSpace(timeStr)) return 0L;

            // Chuẩn hóa: thay comma bằng dot để thống nhất xử lý
            timeStr = timeStr.Trim().Replace(',', '.');

            // Regex: HH:MM:SS.mmm (có thể bỏ HH hoặc MM)
            var match = System.Text.RegularExpressions.Regex.Match(timeStr, @"^(\d+):(\d{2}):(\d{2})\.(\d+)$");
            if (!match.Success)
            {
                return -1L;
            }

            try
            {
                var hours = int.Parse(match.Groups[1].Value);
                var minutes = int.Parse(match.Groups[2].Value);
                var seconds = int.Parse(match.Groups[3].Value);
                var fracSeconds = match.Groups[4].Value;

                // Parse fractional seconds (có thể là 1-3 digits)
                int milliseconds = 0;
                if (fracSeconds.Length == 1)
                {
                    milliseconds = int.Parse(fracSeconds) * 100;
                }
                else if (fracSeconds.Length == 2)
                {
                    milliseconds = int.Parse(fracSeconds) * 10;
                }
                else
                {
                    milliseconds = int.Parse(fracSeconds.Substring(0, 3));
                }

                return (hours * 3600L + minutes * 60L + seconds) * 1000L + milliseconds;
            }
            catch
            {
                return -1L;
            }
        }

        /// <summary>
        /// Parse SRT time string (00:00:00,000) sang milliseconds
        /// </summary>
        private long ParseSrtTimeToMs(string timeStr)
        {
            var parts = timeStr.Split(':', ',');
            if (parts.Length != 4) return 0L;

            var hours = int.Parse(parts[0]);
            var minutes = int.Parse(parts[1]);
            var seconds = int.Parse(parts[2]);
            var milliseconds = int.Parse(parts[3]);

            return hours * 3600000L + minutes * 60000L + seconds * 1000L + milliseconds;
        }

        /// <summary>
        /// Convert milliseconds sang SRT time string (00:00:00,000)
        /// </summary>
        private string MsToSrtTime(long ms)
        {
            var isNegative = ms < 0;
            ms = Math.Abs(ms);

            var hours = ms / 3600000L;
            ms = ms % 3600000L;
            var minutes = ms / 60000L;
            ms = ms % 60000L;
            var seconds = ms / 1000L;
            var milliseconds = ms % 1000L;

            return string.Format("{0:D2}:{1:D2}:{2:D2},{3:D3}", hours, minutes, seconds, milliseconds);
        }

        /// <summary>
        /// Parse ASS time string (H:MM:SS.cc) sang milliseconds
        /// </summary>
        private long ParseAssTimeToMs(string timeStr)
        {
            var parts = timeStr.Split(':', '.');
            if (parts.Length != 4) return 0L;

            var hours = int.Parse(parts[0]);
            var minutes = int.Parse(parts[1]);
            var seconds = int.Parse(parts[2]);
            var centiseconds = int.Parse(parts[3]);

            return hours * 3600000L + minutes * 60000L + seconds * 1000L + centiseconds * 10L;
        }

        /// <summary>
        /// Convert milliseconds sang ASS time string (H:MM:SS.cc)
        /// </summary>
        private string MsToAssTime(long ms)
        {
            var isNegative = ms < 0;
            ms = Math.Abs(ms);

            var hours = ms / 3600000L;
            ms = ms % 3600000L;
            var minutes = ms / 60000L;
            ms = ms % 60000L;
            var seconds = ms / 1000L;
            var centiseconds = (ms % 1000L) / 10L;

            return string.Format("{0}:{1:D2}:{2:D2}.{3:D2}", hours, minutes, seconds, centiseconds);
        }

        #endregion

        #region Karaoke Sync - Toast

        private async void ShowToastSync(string message)
        {
            ToastTextSync.Text = message;
            ToastBorderSync.Visibility = Visibility.Visible;
            await Task.Delay(2000);
            ToastBorderSync.Visibility = Visibility.Collapsed;
        }

        #endregion

    }
}
